File: Lab3/Lab3.cs
namespace ProducerConsumer
{
    class Program
    {
        static void Main(string[] args)
        {
            var program = new Program(3, 150, 3, 12);

            Console.ReadKey();
        }

        private Program(int storageSize, int itemNumber, int producers, int consumers)
        {
            Access = new Semaphore(1, 1);
            Full = new Semaphore(storageSize, storageSize);
            Empty = new Semaphore(0, storageSize);

            for (int i = 0; i < producers - 1; i++)
            {
                (new Thread(Producer)).Start(itemNumber / producers);
            }

            // Leftover
            (new Thread(Producer)).Start(itemNumber - itemNumber / producers * (producers - 1));

            for (int i = 0; i < consumers - 1; i++)
            {
                (new Thread(Consumer)).Start(itemNumber / consumers);
            }

            // Leftover
            (new Thread(Consumer)).Start(itemNumber - itemNumber / consumers * (consumers - 1));
        }

        private readonly Semaphore Access;
        private readonly Semaphore Full;
        private readonly Semaphore Empty;

        private readonly List<string> storage = new();

        private void Producer(object? itemNumber)
        {
            int maxItem = (itemNumber as int?) ?? 0;

            for (int i = 0; i < maxItem; i++)
            {
                Full.WaitOne();
                Thread.Sleep(60);
                Access.WaitOne();

                storage.Add($"item {i} ({Environment.CurrentManagedThreadId})");
                Console.WriteLine($"Producer id {Environment.CurrentManagedThreadId} added item {i}");

                Access.Release();
                Empty.Release();
            }

            Console.WriteLine($"Producer id {Environment.CurrentManagedThreadId} finished");
        }

        private void Consumer(object? itemNumber)
        {
            int maxItem = (itemNumber as int?) ?? 0;

            for (int i = 0; i < maxItem; i++)
            {
                Empty.WaitOne();
                Thread.Sleep(100);
                Access.WaitOne();

                string item = storage.ElementAt(0);
                storage.RemoveAt(0);
                Console.WriteLine($"Consumer id {Environment.CurrentManagedThreadId} removed item {item}");

                //------------------
                //Normaly it must be changed
                Full.Release();

                Access.Release();
                //-------------
            }

            Console.WriteLine($"Consumer id {Environment.CurrentManagedThreadId} finished");
        }
    }
}

File: Lab3/Lab3.java
/* package whatever; // don't place package name! */

import java.util.*;
import java.lang.*;
import java.io.*;
import java.time.*;
import java.util.ArrayList;
import java.util.concurrent.Semaphore;

/* Name of the class has to be "Main" only if the class is public. */
class Ideone
{
	public static void main(String[] args) {
        Ideone main = new Ideone();
        int storageSize = 4;
        int itemNumbers = 40;
        main.starter(storageSize, itemNumbers, 2, 5);
    }

    private void starter(int storageSize, int itemNumbers, int producers, int consumers) {
        Manager manager = new Manager(storageSize);
        
        for(int i = 0; i < producers; i++) {
        	new Producer(i * itemNumbers / producers, (i + 1) * itemNumbers / producers, manager);
        }
        
        for(int i = 0; i < consumers; i++) {
        	new Consumer((i + 1) * itemNumbers / consumers - i * itemNumbers / consumers, manager);
        }
    }
}

class Consumer implements Runnable {
    private final int itemCount;
    private final Manager manager;

    public Consumer(int itemCount, Manager manager) {
        this.itemCount = itemCount;
        this.manager = manager;

        new Thread(this).start();
    }

    @Override
    public void run() {
        for (int i = 0; i < itemCount; i++) {
            String item;
            try {
                manager.empty.acquire();
                Thread.sleep(10);
                manager.access.acquire();

                item = manager.storage.get(0);
                manager.storage.remove(0);
                System.out.println("Consumer " + Thread.currentThread().getId() + " took " + item);

                manager.access.release();
                manager.full.release();

            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

class Producer implements Runnable{
    private final int fromId;
    private final int toId;
    private final Manager manager;

    public Producer(int fromId, int toId, Manager manager) {
        this.fromId = fromId;
        this.toId = toId;
        this.manager = manager;

        new Thread(this).start();
    }

    @Override
    public void run() {
        for (int i = fromId; i < toId; i++) {
            try {
                manager.full.acquire();
                manager.access.acquire();

                manager.storage.add("item " + i);
                System.out.println("Producer " + Thread.currentThread().getId() + " added item " + i);

                manager.access.release();
                manager.empty.release();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

class Manager {

    public Semaphore access;
    public Semaphore full;
    public Semaphore empty;

    public ArrayList<String> storage = new ArrayList<>();

    public Manager(int storageSize) {
        access = new Semaphore(1);
        full = new Semaphore(storageSize);
        empty = new Semaphore(0);
    }
}